Fixes bug where sending email could silently fail

briankendall opened the pull request:
This pull request includes two relatively minor fixes:

1. Fixes a bug with the smtp client where in certain cases sending an email when a torrent completes would silently fail. Cases where it could happen includes but is not limited to: trying to connect to an SMTP server without encryption when it requires SSL, the connecting timing out, or the connection being refused.

2. Hides the "This server requires a secure connection (SSL)" checkbox in the settings when qBittorrent is built without SSL support, as all emails will be unencrypted anyway. This is consistent with the SSL aspects of the Web UI settings being hidden when the build does not include SSL.

glassez commented on src/base/net/smtp.h:
```diff
@@ -39,6 +39,11 @@
+#include <QSslSocket>
+#else
+#include <QTcpSocket>
+#endif
```
Why did you add these includes here?

briankendall commented:
> Why did you add these includes here?

It's necessary for one of the signals I added to Smtp that takes QAbstractSocket::SocketError as a parameter. That said I've changed it to simply include QAbstractSocket.

Chocobo1 commented:
In commit msg:
>Fixes a number of cases where sending an email will silently fail

This is somewhat misleading IMO, you should instead mention "Add log message ..."
And you can squash the commits to 2.

briankendall commented:
> This is somewhat misleading IMO, you should instead mention "Add log message ..."
And you can squash the commits to 2.

Done.

sledgehammer999 commented on src/base/net/smtp.cpp:
```diff
@@ -525,3 +526,12 @@ QString Smtp::getCurrentDateTime() const
+{
+    // Getting a remote host closed error is apparently normal, even when successfully sending
+    // an email
+    if (socketError != QAbstractSocket::RemoteHostClosedError) {
```
As per our [coding guidelines](https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md) remove the curly braces here.

sledgehammer999 commented:
Almost ready for merging but here is a snippet of the [Qt docs](https://doc.qt.io/qt-5/qabstractsocket.html#error-1) for the error() signal of the socket:
>QAbstractSocket::SocketError is not a registered metatype, so for queued connections, you will have to register it with Q_DECLARE_METATYPE() and qRegisterMetaType().

So could you fix it?

PS: We already use those 2 macro/functions in the codebase. Do a search to get an idea on how to do it. If you still can't do it, just remove the curly braces and I'll do it.
PS2: Of course merge the changes into their respective commits. Leave a message here so I get notified of the update.

briankendall commented:
@sledgehammer999 Fixed the curly braces and included Q_REGISTER_METATYPE. It turns out that note in the documentation is incorrect starting with Qt 5.0, as starting then QAbstractSocket::SocketError's metatype is declared in qabstractsocket.h, so I check to make sure the Qt version is older than 5.0.

sledgehammer999 commented on src/base/net/smtp.h:
```diff
@@ -123,4 +125,8 @@ namespace Net
     };
 }
 
+#if QT_VERSION < QT_VERSION_CHECK(5, 0, 0)
```
Change it to `#ifndef QBT_USES_QT5`.

briankendall commented:
Offhandedly I'm not sure why it's failing to build in Qt 4 but I'll fix.

sledgehammer999 commented:
This seems ok now. Thank you for the patch.
